### PR TITLE
Event

### LINKED ISSUE #47: Allow "event:" within types as per JSDoc specs
Hi,

I'm appreciating your library as used by https://github.com/gajus/eslint-plugin-jsdoc ...

I've run into two issues, however, one of which pertains to events...

If I add the event type `module:hurler~event:snowball` (as shown on http://usejsdoc.org/tags-listens.html ), at http://kuniwak.github.io/jsdoctypeparser/ , I get an error. Removing "event:" makes it work.

Per http://usejsdoc.org/tags-event.html , "event:" is allowed in the signature and apparently is even required in some locations:

>  **Syntax**
>
> @event &lt;className>#[event:]&lt;eventName> 
> ...
> JSDoc automatically prepends the namespace event: to each event's name. In general, you must include this namespace when you link to the event in another doclet. (The @fires tag is a notable exception; it allows you to omit the namespace.)

Moreover, the [JSDoc (for @listens)](http://usejsdoc.org/tags-listens.html) states:

> The following example shows how to document an event named module:hurler~event:snowball, as well as a method named module:playground/monitor.reportThrowage that listens for the event...

So "event:" should be permitted within references to events (including `@param`) without complaint, at least at preceding a property name at the beginning of [namepaths](http://usejsdoc.org/about-namepaths.html#namepaths-in-jsdoc-3) (`#`, `.`, or `~` for JSDoc 3, and apparently `-`, if supporting JSDoc 2), so, e.g., `MyClass#event:foo` or `MyClass~event:foo`.

Even if one could work around this by avoiding "event:" within event names, it is not only explicitly permitted by JSDoc but is helpfully descriptive in distinguishing such events from other object references.

Thanks!

(Btw, FWIW, I had reported a similar issue for ESLint's subproject, at https://github.com/eslint/doctrine/issues/221 , but this project is no longer being maintained.)

### PR DESCRIPTION
**Pull request by brettz9**
Builds on PR #48.

- Fix: Allow `event:` prefix within `ModulePathExpr` and `NamepathExpr`; defined at http://usejsdoc.org/tags-event.html#syntax and used within module path at http://usejsdoc.org/tags-listens.html#example (fixes #47)

### CODE REVIEW

File: peg_src/jsdoctype.pegjs [hunk 1]
```diff
@@ -59,10 +59,10 @@ JsIdentifier = $([a-zA-Z_$][a-zA-Z0-9_$]*)
 
 // It is transformed to remove left recursion.
 // See https://en.wikipedia.org/wiki/Left_recursion#Removing_left_recursion
-NamepathExpr = rootOwner:(ParenthesizedExpr / TypeNameExpr) memberPartWithOperators:(_ InfixNamepathOperator _ MemberName)* {
+NamepathExpr = rootOwner:(ParenthesizedExpr / TypeNameExpr) memberPartWithOperators:(_ InfixNamepathOperator _ "event:"? _ MemberName)* {
```
Comment (Kuniwak) [inline peg_src/jsdoctype.pegjs]: Thanks for the contribution!

I think `event:` should be similar to `external:` that defined as [`ExternalNameExpr`](https://github.com/Kuniwak/jsdoctypeparser/blob/8f0bf03b1e658d8a9956f79e069cfdd64c027428/peg_src/jsdoctype.pegjs#L185-L195).  Could you define a new expression that is like `EventNameExpr` instead of extending the `NamepathExpr`?

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Sure, and thanks for the timely response... Will try to take a look to change now...

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: I guess I'm at a loss on how to do this... (And it also would need to be done for `ModulePathExpr`)...

If it is an event name expression, it would seem to imply there should also be non-event name expression. The way I treated it was not to give it its own `NodeType`, though I guess it should probably have its own type. In any case, if it is a new expression, I'm not sure how to swap it in at that point *within* the expression...

It's a bit different from `external:` as this "event:" is not preceding the whole `NamepathExp`, but is just a portion within it... It seems I'd have to reimplement `NamepathExpr` but with the prefix to avoid putting it in `NamepathExpr`. Is that what you'd like?

Comment (Kuniwak) [reply peg_src/jsdoctype.pegjs]: @brettz9 

> It's a bit different from external: as this "event:" is not preceding the whole NamepathExp, but is just a portion within it...

You're right. I had misunderstood the syntax of `event:`...
And now, I feel your implementation approach is almost right. And I feel it is better that we can know whether the member has a event prefix or not. So a new flag indicating it should be placed on `ModulePathExpr`. How do you feel it?

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Very good to include it I think—especially since discovering that per http://usejsdoc.org/tags-event.html:

> JSDoc automatically prepends the namespace event: to each event's name. In general, you must include this namespace when you link to the event in another doclet. (The @fires tag is a notable exception; it allows you to omit the namespace.)

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Plan to prepare now...

Comment (Kuniwak) [reply peg_src/jsdoctype.pegjs]: I think `hasEventPrefix` is bitty better. because the `isEvent` for some event components can be  `false` in another syntax such as JSDuck/ClosureLibrary (because it do not have to have `event:` prefix on another syntax).

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: I've published an update using `hasEventPrefix`, and also added it to the README and to the default publisher...

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Actually just now, I amended again, this time to add a couple publishing tests, one for members (as there had been none) and one for an event member... I plan to stop now though...

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Is there anything you need further on my end?

**Comment by brettz9** [comment]
Would you be able to take another look? I think this should be ready now...

**Comment by brettz9** [comment]
Could I ask for an ETA on when you might be able to take a look at this?

**Comment by brettz9** [comment]
Bump...

**Comment by brettz9** [comment]
I've rebased, but it seems I am not able to push my commit to `master` in `jsdoctypeparser`, despite seeing myself listed on the project (I guess I need to be enabled for the repo specifically?).

I'm also happy to follow a procedure of submitting PRs if that is desired, but with changes requested here by Kuniwak (and with this project creator no longer active), should I submit a new PR?

**Comment by sandersn** [comment]
Hm, I thought org members would be able to commit freely. I made you an org owner instead of just adding you to the contributor list for the repo. I don't think there's much *social* difference. Maybe some *technical* difference.

I'd like to ship 3.1.0 today so I'll go ahead and merge this PR along with your others.
